File: springboot/junit5/src/main/java/com/example/demo/api/ArticleController.java
package com.example.demo.api;

import com.example.demo.model.Article;
import com.example.demo.model.ArticleDto;
import com.example.demo.service.ArticleService;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping(value = "/api/article")
@CrossOrigin(origins = "*")
@AllArgsConstructor
public class ArticleController {

    private final ArticleService articleService;

    @GetMapping
    public ResponseEntity searchAll(Pageable pageable){
        Page<Article> articlePages = articleService.searchAll(pageable);

        return ResponseEntity.ok(articlePages);
    }

    @GetMapping("/{articleIdx}")
    public ResponseEntity searchOne(@PathVariable Long articleIdx){

        Article article = articleService.searchOne(articleIdx);
        ArticleDto.Resp respArticle = ArticleDto.Resp.of(article);

        return ResponseEntity.ok(respArticle);
    }

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    public void register(ArticleDto.CreateReq createReq){

        articleService.register(createReq);
    }

    @PutMapping("/{articleIdx}")
    @ResponseStatus(value = HttpStatus.OK)
    public void modify(@PathVariable Long articleIdx, ArticleDto.ModifyReq modifyReq){

        articleService.modify(articleIdx, modifyReq);
    }

    @DeleteMapping("/{articleIdx}")
    @ResponseStatus(value = HttpStatus.OK)
    public void remove(@PathVariable Long articleIdx){

        articleService.remove(articleIdx);
    }
}

File: springboot/HibernateLock/src/main/kotlin/com/example/hibernatelock/article/controller/ArticleController.kt
package com.example.hibernatelock.article.controller

import com.example.hibernatelock.article.dto.ArticleCreateRequest
import com.example.hibernatelock.article.dto.ArticleDetailDto
import com.example.hibernatelock.article.dto.ArticleDto
import com.example.hibernatelock.article.dto.ArticleUpdateRequest
import com.example.hibernatelock.article.entity.Article
import com.example.hibernatelock.article.entity.ArticleDetail
import com.example.hibernatelock.article.service.ArticleService
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/article")
class ArticleController(
    private val articleService: ArticleService,
) {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(requestParam: ArticleCreateRequest) {

        this.articleService.create(requestParam)
    }

    @GetMapping("/optimistic/{id}")
    fun findByIdWithOptimisticLock(
        @PathVariable id: Long
    ): ArticleDto = ArticleDto.of(this.articleService.findByIdWithOptimisticLock(id))

    @GetMapping("/optimistic-force/{id}")
    fun findByIdWithOptimisticForceLock(
        @PathVariable id: Long
    ): ArticleDto = ArticleDto.of(this.articleService.findByIdWithOptimisticForceLock(id))

    @GetMapping
    fun findAll(): List<Article> = this.articleService.findAll()

    @GetMapping("/detail/optimistic/{id}")
    fun findDetailByIdWithOptimisticLock(
        @PathVariable id: Long
    ): ArticleDetailDto {

        return ArticleDetailDto.of(
            this.articleService.findDetailByIdWithOptimisticLock(id)
        )
    }

    @GetMapping("/detail/optimistic-force/{id}")
    fun findDetailByIdWithOptimisticForceLock(
        @PathVariable id: Long
    ): ArticleDetailDto {

        return ArticleDetailDto.of(
            this.articleService.findDetailByIdWithOptimisticForceLock(id)
        )
    }

    @PutMapping("/detail/optimistic-force/{id}")
    fun updateDetailByIdWithOptimisticForceLock(
        @PathVariable id: Long,
        articleUpdateRequest: ArticleUpdateRequest,
    ): ArticleDetailDto {

        val articleDetail = this.articleService.updateDetailWithOptimisticForceLock(id, articleUpdateRequest)
        return ArticleDetailDto.of(
            articleDetail
        )
    }

    @GetMapping("/shared/{id}")
    fun findByIdWithSharedLock(
        @PathVariable id: Long
    ): ArticleDto = ArticleDto.of(this.articleService.findByIdWithSharedLock(id))

    @GetMapping("/write/{id}")
    fun findByIdWithWriteLock(
        @PathVariable id: Long
    ): ArticleDto = ArticleDto.of(this.articleService.findByIdWithWriteLock(id))

    @GetMapping("/detail/shared/{id}")
    fun findDetailByIdWithSharedLock(
        @PathVariable id: Long
    ): ArticleDetailDto {

        return ArticleDetailDto.of(
            this.articleService.findDetailByIdWithSharedLock(id)
        )
    }

    @GetMapping("/detail/write/{id}")
    fun findDetailByIdWithWriteLock(
        @PathVariable id: Long
    ): ArticleDetailDto {

        return ArticleDetailDto.of(
            this.articleService.findDetailByIdWithWriteLock(id)
        )
    }
}